biomage/utils/data.py

```diff
@@ -0,0 +1,162 @@
+import boto3
+import click
+from botocore.exceptions import ClientError
+
+
+def modified_records(item, target_table, config):
+    """
+    Return modified records.
+    This function should return spreadable dictionary
+    """
+
+    if target_table == config["staging-experiments-table"]:
+        return {
+            # Rewrite pipeline details in experiments-table
+            # metadata to pipeline ARN in production environment
+            "meta": {
+                "M": {
+                    **item["meta"]["M"],
+                    "pipeline": {
+                        "M": {
+                            "stateMachineArn": {"S": ""},
+                            "executionArn": {"S": ""},
+                        }
+                    },
+                }
+            }
+        }
+
+    return {}
+
+
+def same_object(target, source):
+    """
+    Check if a target object is the same with source object by comparing their etags.
+    NOTE: two objects might still be equal and have different etag
+    The method is useful to avoid copying again objects that are definitely
+    equal.
+    """
+    same_etag = False
+
+    try:
+        s3 = boto3.client("s3")
+        s3.head_object(
+            Bucket=target["Bucket"], Key=target["Key"], IfMatch=source["ETag"]
+        )
+        same_etag = True
+    except ClientError as err:
+        # HTTPError 404 is returned if the object doesn't exist
+        # HTTPError 412 is returned if the target's ETag doesn't match with the source's ETag
+        # as a consequence of using `IfMatch` in `s3.head_object()`
+        if err.response["ResponseMetadata"]["HTTPStatusCode"] not in [412, 404]:
+            raise
```

Comment:
if we end up raising an error here (there will neither true, not false returned in that case), then will we still download the s3 object or skip the download?

Reply:
it will actually crash, good catch. I've modified it so exceptions are "ignored" and the file is copied if there's any issue.

Reply:
sounds good